pkg/puller/watcher.go

```diff
@@ -0,0 +1,184 @@
+package puller
+
+import (
+	"encoding/json"
+	"github.com/fsnotify/fsnotify"
+	"github.com/google/go-cmp/cmp"
+	"github.com/kubeflow/kfserving/pkg/apis/serving/v1beta1"
+	"github.com/kubeflow/kfserving/pkg/constants"
+	"github.com/kubeflow/kfserving/pkg/modelconfig"
+	"golang.org/x/sync/syncmap"
+	"io/ioutil"
+	"log"
+	"path/filepath"
+	"time"
+)
+
+var w *Watcher
+
+type Watcher struct {
+	configDir    string
+	ModelTracker *syncmap.Map
+	numWorkers   int
+}
+
+func init() {
+	w = NewWatcher()
+}
+
+func NewWatcher() *Watcher {
+	w := new(Watcher)
+	m := new(syncmap.Map)
+	w.ModelTracker = m
+	return w
+}
+
+type LoadState string
+
+const (
+	// State Related
+	ShouldLoad   LoadState = "Load"
+	ShouldUnload LoadState = "Unload"
+)
+
+type EventWrapper struct {
+	ModelSpec      *v1beta1.ModelSpec
+	LoadState      LoadState
+	ShouldDownload bool
+}
+
+type ModelWrapper struct {
+	ModelSpec *v1beta1.ModelSpec
```

Comment:
use no-pointer type as it is small struct and required field

Reply:
not necessarily, If its an unload command, we only know the modelName, there is no ModelSpec